In [ ]:
# RUN THIS COMMAND ONLY IF YOU USE GOOGLE COLAB.
from google.colab import drive
drive.mount('/content/drive')

In [ ]:
# RUN THIS COMMAND ONLY IF YOU USE GOOGLE COLAB.
%cd drive/MyDrive/TechLabs/02_Data\ Manipulation/Part\ C\ -\ Pandas

# Chapter 08 - Grouping and Aggregating - Analyzing and Exploring Your Data
### Hey Techie,   
Welcome to the eigth notebook of this Pandas tutorial series. We encourage you to take this notebook as a template to code along the instruction video, which you may find at: https://youtu.be/txMdrV1Ut64. In the instruction video, Corey explains how to group and aggregate data using Pandas to extract meaningful insights. As always, at the end of this notebook are practice tasks to test your skills.     

**Here you may find the Pandas documentation:** https://pandas.pydata.org/docs/reference/index.html

#### Have fun! :-)   
    
*Video length*: 50 minutes   
*Self-study time*: 50 minutes   
*Total*: **100 minutes**
<hr style="border:2px solid gray"> </hr>   

## Real-word Example

In [2]:
# This is the convention used to import Pandas.
import pandas as pd

In [2]:
# These options help us to inspect our data more easily.
pd.set_option("display.max_columns", 85)
pd.set_option("display.max_rows", 85)

In [3]:
# These commands load the same survey data Corey is using in his video.
df = pd.read_csv("data/survey_results_public.csv", index_col = "Respondent")
schema_df = pd.read_csv("data/survey_results_schema.csv", index_col = "Column")

In [5]:
# START YOUR CODE HERE.

df.head()


MainBranch Hobbyist  \
Respondent                                                               
1                      I am a student who is learning to code      Yes   
2                      I am a student who is learning to code       No   
3           I am not primarily a developer, but I write co...      Yes   
4                              I am a developer by profession       No   
5                              I am a developer by profession      Yes   

                           OpenSourcer  \
Respondent                               
1                                Never   
2              Less than once per year   
3                                Never   
4                                Never   
5           Once a month or more often   

                                                   OpenSource  \
Respondent                                                      
1           The quality of OSS and closed source software ...   
2           The quality of OSS and closed source software ...   
3           The quality of OSS and closed source software ...   
4           The quality of OSS and closed source software ...   
5           OSS is, on average, of HIGHER quality than pro...   

                                        Employment                 Country  \
Respondent                                                                   
1           Not employed, and not looking for work          United Kingdom   
2               Not employed, but looking for work  Bosnia and Herzegovina   
3                               Employed full-time                Thailand   
4                               Employed full-time           United States   
5                               Employed full-time                 Ukraine   

                   Student                                            EdLevel  \
Respondent                                                                      
1                       No                          Primary/elementary school   
2           Yes, full-time  Secondary school (e.g. American high school, G...   
3                       No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
4                       No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
5                       No           Bachelor’s degree (BA, BS, B.Eng., etc.)   

                                               UndergradMajor  \
Respondent                                                      
1                                                         NaN   
2                                                         NaN   
3                               Web development or web design   
4           Computer science, computer engineering, or sof...   
5           Computer science, computer engineering, or sof...   

                                                     EduOther  \
Respondent                                                      
1           Taught yourself a new language, framework, or ...   
2           Taken an online course in programming or softw...   
3           Taught yourself a new language, framework, or ...   
4           Taken an online course in programming or softw...   
5           Taken an online course in programming or softw...   

                             OrgSize  \
Respondent                             
1                                NaN   
2                                NaN   
3               100 to 499 employees   
4               100 to 499 employees   
5           10,000 or more employees   

                                                      DevType YearsCode  \
Respondent                                                                
1                                                         NaN         4   
2           Developer, desktop or enterprise applications;...       NaN   
3           Designer;Developer, back-end;Developer, front-...         3   
4                                       Developer, full-stack         3   
5           Academic re

In [6]:
df['ConvertedComp'].head(15)

Respondent
1          NaN
2          NaN
3       8820.0
4      61000.0
5          NaN
6     366420.0
7          NaN
8          NaN
9      95179.0
10     13293.0
11         NaN
12         NaN
13     90000.0
14     57060.0
15         NaN
Name: ConvertedComp, dtype: float64

In [7]:
df['ConvertedComp'].median()

57287.0

In [8]:
df.median()

/var/folders/wv/4ykr26xn1xz7lvsm_2btgjc00000gn/T/ipykernel_5493/530051474.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  df.median()


CompTotal        62000.0
ConvertedComp    57287.0
WorkWeekHrs         40.0
CodeRevHrs           4.0
Age                 29.0
dtype: float64

In [10]:
df.describe()

CompTotal  ConvertedComp   WorkWeekHrs    CodeRevHrs           Age
count  5.594500e+04   5.582300e+04  64503.000000  49790.000000  79210.000000
mean   5.519014e+11   1.271107e+05     42.127197      5.084308     30.336699
std    7.331926e+13   2.841523e+05     37.287610      5.513931      9.178390
min    0.000000e+00   0.000000e+00      1.000000      0.000000      1.000000
25%    2.000000e+04   2.577750e+04     40.000000      2.000000     24.000000
50%    6.200000e+04   5.728700e+04     40.000000      4.000000     29.000000
75%    1.200000e+05   1.000000e+05     44.750000      6.000000     35.000000
max    1.000000e+16   2.000000e+06   4850.000000     99.000000     99.000000

In [12]:
df['ConvertedComp'].count()

55823

In [14]:
df['Hobbyist']

Respondent
1        Yes
2         No
3        Yes
4         No
5        Yes
        ... 
88377    Yes
88601     No
88802     No
88816     No
88863    Yes
Name: Hobbyist, Length: 88883, dtype: object

In [16]:
df['Hobbyist'].value_counts()

Yes    71257
No     17626
Name: Hobbyist, dtype: int64

In [19]:
df['SocialMedia']

Respondent
1          Twitter
2        Instagram
3           Reddit
4           Reddit
5         Facebook
           ...    
88377      YouTube
88601          NaN
88802          NaN
88816          NaN
88863     WhatsApp
Name: SocialMedia, Length: 88883, dtype: object

In [20]:
schema_df

QuestionText
Column                                                                   
Respondent              Randomized respondent ID number (not in order ...
MainBranch              Which of the following options best describes ...
Hobbyist                                          Do you code as a hobby?
OpenSourcer                   How often do you contribute to open source?
OpenSource              How do you feel about the quality of open sour...
Employment              Which of the following best describes your cur...
Country                         In which country do you currently reside?
Student                 Are you currently enrolled in a formal, degree...
EdLevel                 Which of the following best describes the high...
UndergradMajor          What was your main or most important field of ...
EduOther                Which of the following types of non-degree edu...
OrgSize                 Approximately how many people are employed by ...
DevType                 Which of the following describe you? Please se...
YearsCode               Including any education, how many years have y...
Age1stCode              At what age did you write your first line of c...
YearsCodePro            How many years have you coded professionally (...
CareerSat               Overall, how satisfied are you with your caree...
JobSat                  How satisfied are you with your current job? (...
MgrIdiot                How confident are you that your manager knows ...
MgrMoney                Do you believe that you need to be a manager t...
MgrWant                 Do you want to become a manager yourself in th...
JobSeek                 Which of the following best describes your cur...
LastHireDate            When was the last time that you took a job wit...
LastInt                 In your most recent successful job interview (...
FizzBuzz                Have you ever been asked to solve FizzBuzz in ...
JobFactors              Imagine that you are deciding between two job ...
ResumeUpdate            Think back to the last time you updated your r...
CurrencySymbol          Which currency do you use day-to-day? If your ...
CurrencyDesc            Which currency do you use day-to-day? If your ...
CompTotal               What is your current total compensation (salar...
CompFreq                 Is that compensation weekly, monthly, or yearly?
ConvertedComp           Salary converted to annual USD salaries using ...
WorkWeekHrs              On average, how many hours per week do you work?
WorkPlan                          How structured or planned is your work?
WorkChallenge           Of these options, what are your greatest chall...
WorkRemote                                How often do you work remotely?
WorkLoc                                   Where would you prefer to work?
ImpSyn                  For the specific work you do, and the years of...
CodeRev                          Do you review code as part of your work?
CodeRevHrs              On average, how many hours per week do you spe...
UnitTests               Does your company regularly employ unit tests ...
PurchaseHow             How does your company make decisions about pur...
PurchaseWhat            What level of influence do you, personally, ha...
LanguageWorkedWith      Which of the following programming, scripting,...
LanguageDesireNextYear  Which of the following programming, scripting,...
DatabaseWorkedWith      Which of the following database environments h...
DatabaseDesireNextYear  Which of the following database environments h...
PlatformWorkedWith      Which of the following platforms have you done...
PlatformDesireNextYear  Which of the following platforms have you done...
WebFrameWorkedWith      Which of the following web frameworks have you...
WebFrameDesireNextYear  Which of the following web frameworks have you...
MiscTechWorkedWith      Which of the following other frameworks, libra...
MiscTechDesireNextYear  Which of the following other frameworks, 

In [21]:
schema_df.loc['SocialMedia']

QuestionText    What social media site do you use the most?
Name: SocialMedia, dtype: object

In [22]:
df['SocialMedia'].value_counts()

Reddit                      14374
YouTube                     13830
WhatsApp                    13347
Facebook                    13178
Twitter                     11398
Instagram                    6261
I don't use social media     5554
LinkedIn                     4501
WeChat 微信                     667
Snapchat                      628
VK ВКонта́кте                 603
Weibo 新浪微博                     56
Youku Tudou 优酷                 21
Hello                          19
Name: SocialMedia, dtype: int64

In [23]:
df['SocialMedia'].value_counts(normalize=True)

Reddit                      0.170233
YouTube                     0.163791
WhatsApp                    0.158071
Facebook                    0.156069
Twitter                     0.134988
Instagram                   0.074150
I don't use social media    0.065777
LinkedIn                    0.053306
WeChat 微信                   0.007899
Snapchat                    0.007437
VK ВКонта́кте               0.007141
Weibo 新浪微博                  0.000663
Youku Tudou 优酷              0.000249
Hello                       0.000225
Name: SocialMedia, dtype: float64

In [24]:
# process

# spliting the object
# applying the function
# combining results

SyntaxError: invalid syntax (1576384442.py, line 3)

In [25]:
df['Country']

Respondent
1                United Kingdom
2        Bosnia and Herzegovina
3                      Thailand
4                 United States
5                       Ukraine
                  ...          
88377                    Canada
88601                       NaN
88802                       NaN
88816                       NaN
88863                     Spain
Name: Country, Length: 88883, dtype: object

In [26]:
df['Country'].value_counts()

United States        20949
India                 9061
Germany               5866
United Kingdom        5737
Canada                3395
                     ...  
Tonga                    1
Timor-Leste              1
North Korea              1
Brunei Darussalam        1
Chad                     1
Name: Country, Length: 179, dtype: int64

In [27]:
df.groupby(['Country'])

In [28]:
country_grp = df.groupby(['Country'])

In [29]:
country_grp

In [30]:
country_grp.get_group('United States')

MainBranch Hobbyist  \
Respondent                                            
4           I am a developer by profession       No   
13          I am a developer by profession      Yes   
22          I am a developer by profession      Yes   
23          I am a developer by profession      Yes   
26          I am a developer by profession      Yes   
...                                    ...      ...   
78292                                  NaN       No   
82717                                  NaN       No   
83397                                  NaN      Yes   
85642                                  NaN       No   
88282                                  NaN      Yes   

                                                  OpenSourcer  \
Respondent                                                      
4                                                       Never   
13          Less than once a month but more than once per ...   
22                                    Less than once per year   
23                                    Less than once per year   
26                                    Less than once per year   
...                                                       ...   
78292                              Once a month or more often   
82717                                 Less than once per year   
83397                                 Less than once per year   
85642                                 Less than once per year   
88282                              Once a month or more often   

                                                   OpenSource  \
Respondent                                                      
4           The quality of OSS and closed source software ...   
13          OSS is, on average, of HIGHER quality than pro...   
22          OSS is, on average, of HIGHER quality than pro...   
23          The quality of OSS and closed source software ...   
26          The quality of OSS and closed source software ...   
...                                                       ...   
78292       OSS is, on average, of HIGHER quality than pro...   
82717       The quality of OSS and closed source software ...   
83397                                                     NaN   
85642       OSS is, on average, of LOWER quality than prop...   
88282       The quality of OSS and closed source software ...   

                                                   Employment        Country  \
Respondent                                                                     
4                                          Employed full-time  United States   
13                                         Employed full-time  United States   
22                                         Employed full-time  United States   
23                                         Employed full-time  United States   
26                                         Employed full-time  United States   
...                                                       ...            ...   
78292       Independent contractor, freelancer, or self-em...  United States   
82717                      Not employed, but looking for work  United States   
83397                      Not employed, but looking for work  United States   
85642       Independent contractor, freelancer, or self-em...  United States   
88282                      Not employed, but looking for work  United States   

           Student                                            EdLevel  \
Respondent                                                              
4               No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
13              No        Master’s degree (MA, MS, M.Eng., MBA, etc.)   
22              No  Some college/university study without earning ...   
23              No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
26              No  Some college/university study without earning ...   
...            ...                                                ...   

In [31]:
country_grp.get_group('India')

MainBranch Hobbyist  \
Respondent                                                    
8                      I code primarily as a hobby      Yes   
10                  I am a developer by profession      Yes   
15          I am a student who is learning to code      Yes   
50                  I am a developer by profession      Yes   
65                  I am a developer by profession      Yes   
...                                            ...      ...   
77339                                          NaN      Yes   
79795                                          NaN      Yes   
83862                                          NaN      Yes   
84299                                          NaN      Yes   
86012                                          NaN      Yes   

                                                  OpenSourcer  \
Respondent                                                      
8                                     Less than once per year   
10                                 Once a month or more often   
15                                                      Never   
50                                 Once a month or more often   
65                                                      Never   
...                                                       ...   
77339                                 Less than once per year   
79795       Less than once a month but more than once per ...   
83862                              Once a month or more often   
84299                                                   Never   
86012                              Once a month or more often   

                                                   OpenSource  \
Respondent                                                      
8           OSS is, on average, of HIGHER quality than pro...   
10          OSS is, on average, of HIGHER quality than pro...   
15          OSS is, on average, of HIGHER quality than pro...   
50          OSS is, on average, of LOWER quality than prop...   
65                                                        NaN   
...                                                       ...   
77339       OSS is, on average, of HIGHER quality than pro...   
79795       OSS is, on average, of HIGHER quality than pro...   
83862       OSS is, on average, of HIGHER quality than pro...   
84299       The quality of OSS and closed source software ...   
86012       OSS is, on average, of HIGHER quality than pro...   

                                    Employment Country         Student  \
Respondent                                                               
8           Not employed, but looking for work   India             NaN   
10                          Employed full-time   India              No   
15          Not employed, but looking for work   India  Yes, full-time   
50                          Employed full-time   India              No   
65                          Employed full-time   India              No   
...                                        ...     ...             ...   
77339                       Employed full-time   India  Yes, full-time   
79795                       Employed full-time   India              No   
83862                       Employed full-time   India  Yes, full-time   
84299                       Employed full-time   India  Yes, full-time   
86012                       Employed full-time   India              No   

                                                      EdLevel  \
Respondent                                                      
8                    Bachelor’s degree (BA, BS, B.Eng., etc.)   
10                Master’s degree (MA, MS, M.Eng., MBA, etc.)   
15          Secondary school (e.g. American high school, G...   
50                   Bachelor’s degree (BA, BS, B.Eng., etc.)   
65                   Bachelor’s degree (BA, BS, B.Eng., etc.)   
...                                                       ...   
77339                Bachelor’s degree (BA, B

In [33]:
filt = df['Country'] == 'United States'
df.loc[filt]

MainBranch Hobbyist  \
Respondent                                            
4           I am a developer by profession       No   
13          I am a developer by profession      Yes   
22          I am a developer by profession      Yes   
23          I am a developer by profession      Yes   
26          I am a developer by profession      Yes   
...                                    ...      ...   
78292                                  NaN       No   
82717                                  NaN       No   
83397                                  NaN      Yes   
85642                                  NaN       No   
88282                                  NaN      Yes   

                                                  OpenSourcer  \
Respondent                                                      
4                                                       Never   
13          Less than once a month but more than once per ...   
22                                    Less than once per year   
23                                    Less than once per year   
26                                    Less than once per year   
...                                                       ...   
78292                              Once a month or more often   
82717                                 Less than once per year   
83397                                 Less than once per year   
85642                                 Less than once per year   
88282                              Once a month or more often   

                                                   OpenSource  \
Respondent                                                      
4           The quality of OSS and closed source software ...   
13          OSS is, on average, of HIGHER quality than pro...   
22          OSS is, on average, of HIGHER quality than pro...   
23          The quality of OSS and closed source software ...   
26          The quality of OSS and closed source software ...   
...                                                       ...   
78292       OSS is, on average, of HIGHER quality than pro...   
82717       The quality of OSS and closed source software ...   
83397                                                     NaN   
85642       OSS is, on average, of LOWER quality than prop...   
88282       The quality of OSS and closed source software ...   

                                                   Employment        Country  \
Respondent                                                                     
4                                          Employed full-time  United States   
13                                         Employed full-time  United States   
22                                         Employed full-time  United States   
23                                         Employed full-time  United States   
26                                         Employed full-time  United States   
...                                                       ...            ...   
78292       Independent contractor, freelancer, or self-em...  United States   
82717                      Not employed, but looking for work  United States   
83397                      Not employed, but looking for work  United States   
85642       Independent contractor, freelancer, or self-em...  United States   
88282                      Not employed, but looking for work  United States   

           Student                                            EdLevel  \
Respondent                                                              
4               No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
13              No        Master’s degree (MA, MS, M.Eng., MBA, etc.)   
22              No  Some college/university study without earning ...   
23              No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
26              No  Some college/university study without earning ...   
...            ...                                                ...   

In [34]:
filt = df['Country'] == 'United States'
df.loc[filt]['SocialMedia'].value_counts()

Reddit                      5700
Twitter                     3468
Facebook                    2844
YouTube                     2463
I don't use social media    1851
Instagram                   1652
LinkedIn                    1020
WhatsApp                     609
Snapchat                     326
WeChat 微信                     93
VK ВКонта́кте                  9
Weibo 新浪微博                     8
Hello                          2
Youku Tudou 优酷                 1
Name: SocialMedia, dtype: int64

In [35]:
filt = df['Country'] == 'India'
df.loc[filt]['SocialMedia'].value_counts()

WhatsApp                    2990
YouTube                     1820
LinkedIn                     955
Facebook                     841
Instagram                    822
Twitter                      542
Reddit                       473
I don't use social media     250
Snapchat                      23
Hello                          5
WeChat 微信                      5
VK ВКонта́кте                  4
Youku Tudou 优酷                 2
Weibo 新浪微博                     1
Name: SocialMedia, dtype: int64

In [36]:
country_grp['SocialMedia'].value_counts()

Country      SocialMedia             
Afghanistan  Facebook                    15
             YouTube                      9
             I don't use social media     6
             WhatsApp                     4
             Instagram                    1
                                         ..
Zimbabwe     Facebook                     3
             YouTube                      3
             Instagram                    2
             LinkedIn                     2
             Reddit                       1
Name: SocialMedia, Length: 1220, dtype: int64

In [39]:
country_grp['SocialMedia'].value_counts().head(100)

Country      SocialMedia             
Afghanistan  Facebook                     15
             YouTube                       9
             I don't use social media      6
             WhatsApp                      4
             Instagram                     1
                                        ... 
Bangladesh   Facebook                    313
             YouTube                     137
             LinkedIn                     62
             Twitter                      19
             I don't use social media     17
Name: SocialMedia, Length: 100, dtype: int64

In [42]:
country_grp.get_group('Philippines')

MainBranch Hobbyist  \
Respondent                                                    
71                  I am a developer by profession      Yes   
940         I am a student who is learning to code      Yes   
1003                   I code primarily as a hobby      Yes   
1216                I am a developer by profession      Yes   
1338                I am a developer by profession      Yes   
...                                            ...      ...   
88677               I am a developer by profession      Yes   
27544                                          NaN      Yes   
29998                                          NaN      Yes   
44923                                          NaN       No   
76444                                          NaN       No   

                                                  OpenSourcer  \
Respondent                                                      
71                                                      Never   
940                                   Less than once per year   
1003                                                    Never   
1216        Less than once a month but more than once per ...   
1338                                                    Never   
...                                                       ...   
88677                                                   Never   
27544                              Once a month or more often   
29998       Less than once a month but more than once per ...   
44923                              Once a month or more often   
76444                              Once a month or more often   

                                                   OpenSource  \
Respondent                                                      
71          The quality of OSS and closed source software ...   
940         OSS is, on average, of HIGHER quality than pro...   
1003        OSS is, on average, of HIGHER quality than pro...   
1216        OSS is, on average, of LOWER quality than prop...   
1338        OSS is, on average, of HIGHER quality than pro...   
...                                                       ...   
88677       The quality of OSS and closed source software ...   
27544       OSS is, on average, of LOWER quality than prop...   
29998       OSS is, on average, of LOWER quality than prop...   
44923       OSS is, on average, of HIGHER quality than pro...   
76444       OSS is, on average, of LOWER quality than prop...   

                                                   Employment      Country  \
Respondent                                                                   
71                                         Employed full-time  Philippines   
940                                                       NaN  Philippines   
1003                       Not employed, but looking for work  Philippines   
1216                                       Employed full-time  Philippines   
1338                                       Employed full-time  Philippines   
...                                                       ...          ...   
88677                                      Employed full-time  Philippines   
27544                                      Employed part-time  Philippines   
29998                                      Employed full-time  Philippines   
44923       Independent contractor, freelancer, or self-em...  Philippines   
76444                                      Employed full-time  Philippines   

                   Student                                            EdLevel  \
Respondent                                                                      
71                      No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
940         Yes, part-time  Some college/university study without earning ...   
1003        Yes, full-time  Secondary school (e.g. American high school, G...   
1216                    No  Some college/university study without earning ...   
1338                    No 

In [45]:
country_grp.get_group('Philippines')['SocialMedia'].value_counts().head(50)

Facebook                    200
YouTube                      72
Reddit                       37
Twitter                      25
Instagram                    12
I don't use social media     10
LinkedIn                      8
WhatsApp                      4
Hello                         1
WeChat 微信                     1
Snapchat                      1
Name: SocialMedia, dtype: int64

In [44]:
country_grp['SocialMedia'].value_counts().loc['Philippines']

SocialMedia
Facebook                    200
YouTube                      72
Reddit                       37
Twitter                      25
Instagram                    12
I don't use social media     10
LinkedIn                      8
WhatsApp                      4
Hello                         1
Snapchat                      1
WeChat 微信                     1
Name: SocialMedia, dtype: int64

In [46]:
country_grp['SocialMedia'].value_counts().loc['Germany']

SocialMedia
WhatsApp                    1343
YouTube                      970
Reddit                       923
Twitter                      866
I don't use social media     501
Facebook                     404
Instagram                    264
LinkedIn                     141
Snapchat                      10
VK ВКонта́кте                  8
WeChat 微信                      8
Name: SocialMedia, dtype: int64

In [47]:
country_grp['SocialMedia'].value_counts().loc['China']

SocialMedia
WeChat 微信                   403
YouTube                      53
Weibo 新浪微博                   42
I don't use social media     27
Twitter                      27
Reddit                       12
LinkedIn                     11
Facebook                      8
Instagram                     7
Youku Tudou 优酷                7
WhatsApp                      3
VK ВКонта́кте                 1
Name: SocialMedia, dtype: int64

In [48]:
country_grp['SocialMedia'].value_counts().loc['Russian Federation']

SocialMedia
VK ВКонта́кте               442
YouTube                     302
I don't use social media    155
Reddit                      140
Twitter                     130
Instagram                    94
WhatsApp                     90
Facebook                     67
LinkedIn                     35
Snapchat                      1
Name: SocialMedia, dtype: int64

In [49]:
country_grp['SocialMedia'].value_counts(normalize=True).loc['Russian Federation']

SocialMedia
VK ВКонта́кте               0.303571
YouTube                     0.207418
I don't use social media    0.106456
Reddit                      0.096154
Twitter                     0.089286
Instagram                   0.064560
WhatsApp                    0.061813
Facebook                    0.046016
LinkedIn                    0.024038
Snapchat                    0.000687
Name: SocialMedia, dtype: float64

In [50]:
country_grp['ConvertedComp'].median()

Country
Afghanistan                               6222.0
Albania                                  10818.0
Algeria                                   7878.0
Andorra                                 160931.0
Angola                                    7764.0
                                          ...   
Venezuela, Bolivarian Republic of...      6384.0
Viet Nam                                 11892.0
Yemen                                    11940.0
Zambia                                    5040.0
Zimbabwe                                 19200.0
Name: ConvertedComp, Length: 179, dtype: float64

In [51]:
country_grp['ConvertedComp'].median().loc['Germany']

63016.0

In [52]:
country_grp['ConvertedComp'].agg(['median', 'mean'])

median           mean
Country                                                      
Afghanistan                             6222.0  101953.333333
Albania                                10818.0   21833.700000
Algeria                                 7878.0   34924.047619
Andorra                               160931.0  160931.000000
Angola                                  7764.0    7764.000000
...                                        ...            ...
Venezuela, Bolivarian Republic of...    6384.0   14581.627907
Viet Nam                               11892.0   17233.436782
Yemen                                  11940.0   16909.166667
Zambia                                  5040.0   10075.375000
Zimbabwe                               19200.0   34046.666667

[179 rows x 2 columns]

In [53]:
country_grp['ConvertedComp'].agg(['median', 'mean']).loc['Germany']

median     63016.000000
mean      109256.884066
Name: Germany, dtype: float64

In [54]:
country_grp['ConvertedComp'].agg(['median', 'mean']).loc['Canada']

median     68705.000000
mean      134018.564909
Name: Canada, dtype: float64

In [55]:
filt = df['Country'] == 'India'
df.loc[filt]

MainBranch Hobbyist  \
Respondent                                                    
8                      I code primarily as a hobby      Yes   
10                  I am a developer by profession      Yes   
15          I am a student who is learning to code      Yes   
50                  I am a developer by profession      Yes   
65                  I am a developer by profession      Yes   
...                                            ...      ...   
77339                                          NaN      Yes   
79795                                          NaN      Yes   
83862                                          NaN      Yes   
84299                                          NaN      Yes   
86012                                          NaN      Yes   

                                                  OpenSourcer  \
Respondent                                                      
8                                     Less than once per year   
10                                 Once a month or more often   
15                                                      Never   
50                                 Once a month or more often   
65                                                      Never   
...                                                       ...   
77339                                 Less than once per year   
79795       Less than once a month but more than once per ...   
83862                              Once a month or more often   
84299                                                   Never   
86012                              Once a month or more often   

                                                   OpenSource  \
Respondent                                                      
8           OSS is, on average, of HIGHER quality than pro...   
10          OSS is, on average, of HIGHER quality than pro...   
15          OSS is, on average, of HIGHER quality than pro...   
50          OSS is, on average, of LOWER quality than prop...   
65                                                        NaN   
...                                                       ...   
77339       OSS is, on average, of HIGHER quality than pro...   
79795       OSS is, on average, of HIGHER quality than pro...   
83862       OSS is, on average, of HIGHER quality than pro...   
84299       The quality of OSS and closed source software ...   
86012       OSS is, on average, of HIGHER quality than pro...   

                                    Employment Country         Student  \
Respondent                                                               
8           Not employed, but looking for work   India             NaN   
10                          Employed full-time   India              No   
15          Not employed, but looking for work   India  Yes, full-time   
50                          Employed full-time   India              No   
65                          Employed full-time   India              No   
...                                        ...     ...             ...   
77339                       Employed full-time   India  Yes, full-time   
79795                       Employed full-time   India              No   
83862                       Employed full-time   India  Yes, full-time   
84299                       Employed full-time   India  Yes, full-time   
86012                       Employed full-time   India              No   

                                                      EdLevel  \
Respondent                                                      
8                    Bachelor’s degree (BA, BS, B.Eng., etc.)   
10                Master’s degree (MA, MS, M.Eng., MBA, etc.)   
15          Secondary school (e.g. American high school, G...   
50                   Bachelor’s degree (BA, BS, B.Eng., etc.)   
65                   Bachelor’s degree (BA, BS, B.Eng., etc.)   
...                                                       ...   
77339                Bachelor’s degree (BA, B

In [56]:
filt = df['Country'] == 'India'
df.loc[filt]['LanguageWorkedWith']

Respondent
8        Bash/Shell/PowerShell;C;C++;HTML/CSS;Java;Java...
10                           C#;Go;JavaScript;Python;R;SQL
15       Assembly;Bash/Shell/PowerShell;C;C++;HTML/CSS;...
50       Bash/Shell/PowerShell;C;C++;HTML/CSS;Java;Java...
65                         Assembly;C;C++;C#;HTML/CSS;Java
                               ...                        
77339                                            Other(s):
79795                 Bash/Shell/PowerShell;Python;SQL;VBA
83862    Assembly;C;C++;HTML/CSS;Java;JavaScript;Object...
84299     HTML/CSS;Java;JavaScript;Python;Swift;TypeScript
86012        Bash/Shell/PowerShell;C++;HTML/CSS;JavaScript
Name: LanguageWorkedWith, Length: 9061, dtype: object

In [58]:
filt = df['Country'] == 'India'
df.loc[filt]['LanguageWorkedWith'].str.contains('Python')

Respondent
8         True
10        True
15       False
50        True
65       False
         ...  
77339    False
79795     True
83862    False
84299     True
86012    False
Name: LanguageWorkedWith, Length: 9061, dtype: object

In [59]:
filt = df['Country'] == 'India'
df.loc[filt]['LanguageWorkedWith'].str.contains('Python').sum()

3105

In [60]:
country_grp['LanguageWorkedWith'].str.contains('Python').sum()

AttributeError: 'SeriesGroupBy' object has no attribute 'str'

In [61]:
country_grp['LanguageWorkedWith'].apply(lambda series: series.str.contains('Python').sum())

Country
Afghanistan                              8
Albania                                 23
Algeria                                 40
Andorra                                  0
Angola                                   2
                                        ..
Venezuela, Bolivarian Republic of...    28
Viet Nam                                78
Yemen                                    3
Zambia                                   4
Zimbabwe                                14
Name: LanguageWorkedWith, Length: 179, dtype: int64

In [64]:
# What % of people from each county know Python?

# trial answer

country_grp.apply(lambda series: series.str.contains('Python').sum()).value_counts(normalize=True)

AttributeError: 'DataFrame' object has no attribute 'str'

In [66]:
country_respondents = df['Country'].value_counts()
country_respondents

United States        20949
India                 9061
Germany               5866
United Kingdom        5737
Canada                3395
                     ...  
Tonga                    1
Timor-Leste              1
North Korea              1
Brunei Darussalam        1
Chad                     1
Name: Country, Length: 179, dtype: int64

In [67]:
country_uses_python = country_grp['LanguageWorkedWith'].apply(lambda series: series.str.contains('Python').sum())
country_uses_python

Country
Afghanistan                              8
Albania                                 23
Algeria                                 40
Andorra                                  0
Angola                                   2
                                        ..
Venezuela, Bolivarian Republic of...    28
Viet Nam                                78
Yemen                                    3
Zambia                                   4
Zimbabwe                                14
Name: LanguageWorkedWith, Length: 179, dtype: int64

In [69]:
python_df = pd.concat([country_respondents, country_uses_python])
python_df

United States                           20949
India                                    9061
Germany                                  5866
United Kingdom                           5737
Canada                                   3395
                                        ...  
Venezuela, Bolivarian Republic of...       28
Viet Nam                                   78
Yemen                                       3
Zambia                                      4
Zimbabwe                                   14
Length: 358, dtype: int64

In [71]:
python_df = pd.concat([country_respondents, country_uses_python], axis='columns')

In [72]:
python_df

Country  LanguageWorkedWith
United States        20949               10083
India                 9061                3105
Germany               5866                2451
United Kingdom        5737                2384
Canada                3395                1558
...                    ...                 ...
Tonga                    1                   0
Timor-Leste              1                   1
North Korea              1                   0
Brunei Darussalam        1                   0
Chad                     1                   0

[179 rows x 2 columns]

In [73]:
python_df = pd.concat([country_respondents, country_uses_python], axis='columns', sort=False)

In [74]:
python_df

Country  LanguageWorkedWith
United States        20949               10083
India                 9061                3105
Germany               5866                2451
United Kingdom        5737                2384
Canada                3395                1558
...                    ...                 ...
Tonga                    1                   0
Timor-Leste              1                   1
North Korea              1                   0
Brunei Darussalam        1                   0
Chad                     1                   0

[179 rows x 2 columns]

In [75]:
python_df.rename(columns={'Country': 'NumberOfRespondents', 'LanguageWorkedWith': "NumberOfRespondentsKnowsPython"})

NumberOfRespondents  NumberOfRespondentsKnowsPython
United States                    20949                           10083
India                             9061                            3105
Germany                           5866                            2451
United Kingdom                    5737                            2384
Canada                            3395                            1558
...                                ...                             ...
Tonga                                1                               0
Timor-Leste                          1                               1
North Korea                          1                               0
Brunei Darussalam                    1                               0
Chad                                 1                               0

[179 rows x 2 columns]

In [76]:
python_df.rename(columns={'Country': 'NumberOfRespondents', 'LanguageWorkedWith': "NumberOfRespondentsKnowsPython"}, inplace=True)

In [77]:
python_df

NumberOfRespondents  NumberOfRespondentsKnowsPython
United States                    20949                           10083
India                             9061                            3105
Germany                           5866                            2451
United Kingdom                    5737                            2384
Canada                            3395                            1558
...                                ...                             ...
Tonga                                1                               0
Timor-Leste                          1                               1
North Korea                          1                               0
Brunei Darussalam                    1                               0
Chad                                 1                               0

[179 rows x 2 columns]

In [80]:
python_df['PercentageKnowsPythonPerCountry'] = (python_df['NumberOfRespondentsKnowsPython'] / python_df['NumberOfRespondents'] * 100)

In [81]:
python_df

NumberOfRespondents  NumberOfRespondentsKnowsPython  \
United States                    20949                           10083   
India                             9061                            3105   
Germany                           5866                            2451   
United Kingdom                    5737                            2384   
Canada                            3395                            1558   
...                                ...                             ...   
Tonga                                1                               0   
Timor-Leste                          1                               1   
North Korea                          1                               0   
Brunei Darussalam                    1                               0   
Chad                                 1                               0   

                   KnowsPythonPerCountry  PercentageKnowsPythonPerCountry  
United States                  48.131176                        48.131176  
India                          34.267741                        34.267741  
Germany                        41.783157                        41.783157  
United Kingdom                 41.554820                        41.554820  
Canada                         45.891016                        45.891016  
...                                  ...                              ...  
Tonga                           0.000000                         0.000000  
Timor-Leste                   100.000000                       100.000000  
North Korea                     0.000000                         0.000000  
Brunei Darussalam               0.000000                         0.000000  
Chad                            0.000000                         0.000000  

[179 rows x 4 columns]

In [82]:
python_df.sort_values(by='PercentageKnowsPythonPerCountry', ascending=False)

NumberOfRespondents  \
Sao Tome and Principe                               1   
Timor-Leste                                         1   
Dominica                                            1   
Niger                                               1   
Turkmenistan                                        7   
...                                               ...   
Cape Verde                                          3   
Lao People's Democratic Republic                    3   
Malawi                                              2   
Liberia                                             2   
Chad                                                1   

                                  NumberOfRespondentsKnowsPython  \
Sao Tome and Principe                                          1   
Timor-Leste                                                    1   
Dominica                                                       1   
Niger                                                          1   
Turkmenistan                                                   6   
...                                                          ...   
Cape Verde                                                     0   
Lao People's Democratic Republic                               0   
Malawi                                                         0   
Liberia                                                        0   
Chad                                                           0   

                                  KnowsPythonPerCountry  \
Sao Tome and Principe                        100.000000   
Timor-Leste                                  100.000000   
Dominica                                     100.000000   
Niger                                        100.000000   
Turkmenistan                                  85.714286   
...                                                 ...   
Cape Verde                                     0.000000   
Lao People's Democratic Republic               0.000000   
Malawi                                         0.000000   
Liberia                                        0.000000   
Chad                                           0.000000   

                                  PercentageKnowsPythonPerCountry  
Sao Tome and Principe                                  100.000000  
Timor-Leste                                            100.000000  
Dominica                                               100.000000  
Niger                                                  100.000000  
Turkmenistan                                            85.714286  
...                                                           ...  
Cape Verde                                               0.000000  
Lao People's Democratic Republic                         0.000000  
Malawi                                                   0.000000  
Liberia                                                  0.000000  
Chad                                                     0.000000  

[179 rows x 4 columns]

In [83]:
python_df.head(50)

NumberOfRespondents  NumberOfRespondentsKnowsPython  \
United States                     20949                           10083   
India                              9061                            3105   
Germany                            5866                            2451   
United Kingdom                     5737                            2384   
Canada                             3395                            1558   
France                             2391                            1054   
Brazil                             1948                             767   
Poland                             1922                             751   
Australia                          1903                             790   
Netherlands                        1852                             767   
Russian Federation                 1694                             708   
Spain                              1604                             641   
Italy                              1576                             625   
Sweden                             1274                             496   
Switzerland                         978                             460   
Israel                              952                             457   
Turkey                              949                             371   
Pakistan                            923                             251   
Ukraine                             868                             246   
Austria                             839                             338   
Czech Republic                      764                             285   
Romania                             760                             253   
Iran                                738                             268   
Belgium                             727                             296   
China                               664                             297   
Bulgaria                            659                             181   
Mexico                              642                             245   
South Africa                        627                             202   
Denmark                             617                             246   
Bangladesh                          605                             204   
Norway                              574                             232   
Greece                              556                             226   
Argentina                           553                             191   
Finland                             546                             266   
Portugal                            525                             182   
New Zealand                         524                             236   
Nigeria                             522                             152   
Hungary                             513                             191   
Indonesia                           507                             148   
Ireland                             501                             209   
Serbia                              402                             117   
Japan                               391                             182   
Philippines                         381                             117   
Sri Lanka                           372                             137   
Egypt                               323                             111   
Colombia                            313                             108   
Singapore                           303                             139   
Malaysia                            299                              95   
Slovenia                            298                             119   
Slovakia                            260                             101   

                    KnowsPythonPerCountry  PercentageKnowsPythonPerCountry  
United States                   48.131176                        48.131176  
India                           34.26774

In [85]:
python_df.loc['Japan']

NumberOfRespondents                391.000000
NumberOfRespondentsKnowsPython     182.000000
KnowsPythonPerCountry               46.547315
PercentageKnowsPythonPerCountry     46.547315
Name: Japan, dtype: float64

In [86]:
python_df.loc['Brazil']

NumberOfRespondents                1948.000000
NumberOfRespondentsKnowsPython      767.000000
KnowsPythonPerCountry                39.373717
PercentageKnowsPythonPerCountry      39.373717
Name: Brazil, dtype: float64

<hr style="border:2px solid gray"> </hr>   
   
## Practice Tasks   

In [3]:
# Load a clean DataFrame.
df = pd.read_csv("data/survey_results_public.csv", index_col = "Respondent")

#### 1. Create an overview of the coding experience ("YearsCode" column; mean) and the compensation ("ConvertedComp"; median) per country. To do so, change the data type from the "YearsCode" column to float (replace "Less than 1 year" with 0 and "More than 50 years" with 51). Please return a DataFrame named *country_overview* that looks the following:    
| Country     |   YearsCode |   ConvertedComp |
|:------------|------------:|----------------:|
| Afghanistan |     7.31707 |            6222 |
| Albania     |     6.51807 |           10818 |
| ...         |     ...     |            ...  |

   
   
<br /> 
<details>    
<summary>
    <font size="3" color="red"><b>Hints (click to expand)</b></font>
</summary>
<p>
    <ul>
        <li>First, replace the "less than" and "more than" statements within the YearsCode column. This way, you can set the data type of this column to float.</li>
        <li>Group by the countries column.</li>
        <li>To apply different aggregation methods to different columns, pass a dictionary to the agg method. The dictionary's keys correspond to the column names you want to aggregate, and the values indicate the specific aggregation method. For example, groupby_object.agg({"Column 1": "mean", "Column 2": "median"}).</li>
    </ul>
</p>
</details>

In [99]:
# # START YOUR CODE HERE.

# #converting into float
# df["YearsCode"].replace("Less than 1 year", 0)
# df["YearsCode"].replace("More than 50 years", 51)
# df["YearsCode"] = df["YearsCode"].astype(float)

# # grouping by country
# country_grp = df.groupby('Country')

# # df['ConvertedComp'].median()
# # df['YearsCode'].mean()


# country_overview = country_grp.agg({"YearsCode": "mean", "ConvertedComp": "median"})

# country_overview
# # END YOUR CODE HERE.


# solution

df["YearsCode"].replace("Less than 1 year", 0, inplace = True)
df["YearsCode"].replace("More than 50 years", 51, inplace = True)
df["YearsCode"] = df["YearsCode"].astype(float)
groupby_object = df.groupby("Country")
country_overview = groupby_object.agg({"YearsCode": "mean", "ConvertedComp": "median"})

country_overview.rename(columns={'YearsCode': 'YearsCodeMean', 'ConvertedComp': 'ConvertedCompMedian'})

country_overview

YearsCode  ConvertedComp
Country                                                       
Afghanistan                            7.317073         6222.0
Albania                                6.518072        10818.0
Algeria                                6.473684         7878.0
Andorra                               10.500000       160931.0
Angola                                 2.250000         7764.0
...                                         ...            ...
Venezuela, Bolivarian Republic of...   9.477273         6384.0
Viet Nam                               6.504386        11892.0
Yemen                                  5.789474        11940.0
Zambia                                 4.916667         5040.0
Zimbabwe                               6.000000        19200.0

[179 rows x 2 columns]

In [97]:
# THIS CELL TESTS YOUR RESULTS.
assert type(country_overview) == type(pd.DataFrame([])), "Please return a DataFrame!"
assert country_overview.loc["Germany", "YearsCode"] == 12.784108460614382, "Your results seem to be incorrect!"
assert country_overview.loc["United States", "ConvertedComp"] == 110000.0, "Your results seem to be incorrect!"
assert country_overview.loc["India", "ConvertedComp"] == 10080.0, "Your results seem to be incorrect!"

<details>    
<summary>
    <font size="3" color="darkgreen"><b>Solution (click to expand)</b></font>
</summary>
<p>
    <code>df["YearsCode"].replace("Less than 1 year", 0, inplace = True)</code><br />
    <code>df["YearsCode"].replace("More than 50 years", 51, inplace = True)</code><br />
    <code>df["YearsCode"] = df["YearsCode"].astype(float)</code><br />
    <code>groupby_object = df.groupby("Country")</code><br />
    <code>country_overview = groupby_object.agg({"YearsCode": "mean", "ConvertedComp": "median"})</code><br />
</p>
</details>   
   
#### 2. Create an overview that shows what percentage of respondents are very satisfied ("JobSat" column) with their current job per country ("Country" column). Return a Series object named *satisfied_percentage* that is sorted in descending order and looks the following:   
|             |          |
|:------------|---------:|
| Timor-Leste | 1        |
| Belize      | 0.666667 |
| ...         | ...      |

<br /> 
<details>    
<summary>
    <font size="3" color="red"><b>Hints (click to expand)</b></font>
</summary>
<p>
    <ul>
        <li>Group by the country.</li>
        <li>Then determine per country how often a participant answered "Very Satisfied" to the job satisfaction question.</li>
        <li>Determine how many participants there was per country. Note that you previously excluded all participants who did not answer both questions.
Arithmetic operations in Pandas are always performed so that the same indices are offset against each other. Your groupby-DataFrame and the CountryCount-DataFrame both have the countries as the index.</li>
    </ul>
</p>
</details>

In [107]:
# # This drops every participant that did not answer both questions.
# task_df = df[["JobSat", "Country"]].dropna()
# # START YOUR CODE HERE.

# satisfied_percentage = task_df.sort_values(by="JobSat", ascending=False)
# satisfied_percentage
# # END YOUR CODE HERE.

# solution

task_df = df[["JobSat", "Country"]].dropna()
groupby_object = task_df.groupby("Country")
satisfied_per_country = groupby_object["JobSat"].apply(lambda x: (x == "Very satisfied").sum())
respondents_per_country = df["Country"].value_counts()
satisfied_percentage = (satisfied_per_country/respondents_per_country).sort_values(ascending = False)

satisfied_percentage

Timor-Leste              1.000000
Belize                   0.666667
San Marino               0.500000
Bhutan                   0.500000
Fiji                     0.400000
                           ...   
North Korea                   NaN
Papua New Guinea              NaN
Saint Kitts and Nevis         NaN
Sao Tome and Principe         NaN
Tonga                         NaN
Length: 179, dtype: float64

In [103]:
# THIS CELL TESTS YOUR RESULTS.
assert type(satisfied_percentage) == type(pd.Series([], dtype = "int64")), "Please return a Series!"
assert satisfied_percentage.loc["Germany"] == 0.24616433685646097, "Your results seem to be incorrect!"
assert satisfied_percentage.loc["United States"] == 0.3435486180724617, "Your results seem to be incorrect!"
assert satisfied_percentage.loc["India"] == 0.16410992164220284, "Your results seem to be incorrect!"

AssertionError: Please return a Series!

<details>    
<summary>
    <font size="3" color="darkgreen"><b>Solution (click to expand)</b></font>
</summary>
<p>
    <code>task_df = df[["JobSat", "Country"]].dropna()</code><br />
    <code>groupby_object = task_df.groupby("Country")</code><br />
    <code>satisfied_per_country = groupby_object["JobSat"].apply(lambda x: (x == "Very satisfied").sum())</code><br />
    <code>respondents_per_country = df["Country"].value_counts()</code><br />
    <code>satisfied_percentage = (satisfied_per_country/respondents_per_country).sort_values(ascending = False)</code><br />
</p>
</details>   
   
#### 3. The mean is known to be strongly influenced by outliers. Therefore, create an overview showing how many participants per country receive a higher compensation (ConvertedComp column) than the respective country average. Contrast this column with how many people from each country (Country column) generally participated in the study. The first line of code is already given to include only participants that answered both questions. Please return a DataFrame named *comp_average* that looks like this:   
| Country     |   Number of respondents with an above average compensation |   Total number of respondents |
|:------------|-----------------------------------------------------------:|------------------------------:|
| Afghanistan |                                                          2 |                            12 |
| Albania     |                                                         14 |                            50 |
| ...         |                                                        ... |                           ... |   
   

<br />
<details>    
<summary>
    <font size="3" color="red"><b>Hints (click to expand)</b></font>
</summary>
<p>
    <ul>
        <li>There are several ways how to solve this task.</li>
        <li>First, in any case, one must calculate the average compensation per country with a groupby operation.</li>
        <li>Second, one should check per participant whether their compensation is above or below the mean compensation of their country. The resulting boolean values should be stored in a new column in the DataFrame.</li>
        <li>Then one can perform a final groupby operation on the DataFrame by the countries. For the just determined above_average column, one determines the sum (because boolean values are handled like 0 and 1), and for the country column, the count.</li>
    </ul>
        
</p>
</details>

In [ ]:
# This drops every participant that did not answer both questions.
comp_average = df[["Country", "ConvertedComp"]].dropna()
# START YOUR CODE HERE.



# END YOUR CODE HERE.

In [ ]:
# THIS CELL TESTS YOUR RESULTS.
import numpy as np
assert type(comp_average) == type(pd.DataFrame([])), "Please return a DataFrame!"
assert comp_average.shape == (161, 2), "Your DataFrame seems to have the wrong dimensions!"
assert np.array_equal(comp_average.loc["United States"].to_numpy(), [1911, 14981]), "Your results seem to be wrong!"
assert np.array_equal(comp_average.loc["Germany"].to_numpy(), [466, 3778]), "Your results seem to be wrong!"
assert np.array_equal(comp_average.loc["India"].to_numpy(), [677, 3999]), "Your results seem to be wrong!"

<details>    
<summary>
    <font size="3" color="darkgreen"><b>Solution (click to expand)</b></font>
</summary>
<p>
    <code>comp_average = df[["Country", "ConvertedComp"]].dropna()</code><br />
    <code>country_mean = df.groupby("Country").agg({"ConvertedComp": "mean"})</code><br />
    <code>comp_average["above_average"] = [compensation > country_mean.loc[country, "ConvertedComp"]</code><br />
    <code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;for compensation, country in zip(comp_average["ConvertedComp"].values,</code><br /> 
    <code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;comp_average["Country"].values)]</code><br />
    <code>comp_average = comp_average.groupby(</code><br />
    <code>&nbsp;&nbsp;&nbsp;&nbsp;"Country").agg({"above_average": "sum", "Country": "count"}).rename(</code><br />
    <code>&nbsp;&nbsp;&nbsp;&nbsp;{"above_average": "Number of respondents with an above average compensation",</code><br />
    <code>&nbsp;&nbsp;&nbsp;&nbsp; "Country": "Total number of respondents"}, axis = 1)</code><br />
</p>
</details>